### PR TITLE
Change Ir.fun_def to be a record instead of a tuple 

### PR DESCRIPTION
**Pull request by frank-emrich**
This changes the type `Ir.fun_def` from being a 7-tuple to a record. A tedious, but completely mechanical change. As a pleasant side-effect, this makes the output of `show_compiled_ir` much more readable!

### CODE REVIEW

File: core/closures.ml [hunk 1]
```diff
@@ -363,8 +376,8 @@ struct
                treat them as globals *)
           let o, _ =
             List.fold_right
-              (fun (f, _, _, _, _) (o, fs) ->
-                 let o, f = o#binder f in
+              (fun fundef (o, fs) ->
```
Comment (dhil) [inline core/closures.ml:379]: Maybe a bit late now, but for future for these sort of projections I think it is nicer to perform them by pattern matching; in this instance `{ binder; _ }`.

File: core/ir.mli [hunk 2]
```diff
@@ -59,7 +59,16 @@ and tail_computation =
   | Special    of special
   | Case       of value * (binder * computation) name_map * (binder * computation) option
   | If         of value * computation * computation
-and fun_def = binder * (tyvar list * binder list * computation) * binder option * location * bool
+and fun_def =
+  {
+    binder   : binder;
```
Comment (dhil) [inline core/ir.mli:64]: Ideally each of these field names would be prefixed, e.g. `fn_binder : binder`.

File: core/closures.ml (hunk 1 above)
Comment (frank-emrich) [reply core/closures.ml:379]: I've done that in most other cases

File: core/ir.mli (hunk 2 above)
Comment (frank-emrich) [reply core/ir.mli:64]: Unfortunately, these prefixes make record field punning less convenient: For example, we always use `tyvars` as the corresponding variable name, which would require `{fn_tyvar = tyvar; ... }` every single time.

Comment (frank-emrich) [reply core/ir.mli:64]: Do you (or others) feel strongly enough about this so I should change this or not?

Comment (jamescheney) [reply core/ir.mli:64]: I think the reason for such prefixing is that in OCaml there's no way to disambiguate between field names defined in different modules that happen to both be in the same scope: whoever was declared last wins.  Hence, prefixing is a way of defensive programming around this limitation.  (I had a vague recollection that something may have been done about this in modern OCamls but can't remember how/what.)  In any case, I think it's probably better to stick with this convention uniformly even though it makes code a little more painful.

Comment (frank-emrich) [reply core/ir.mli]: I've update the PR to prefix all the fields with `fn_`

**Comment by frank-emrich** [comment]
I assume that people are okay with this and will merge it tomorrow if nobody objects.

**Pull request merged**